### DISCUSSION TITLE
IO of Multiple Types Error

### DISCUSSION THREAD

**Discussion by mrberman87** [SpinalHDL/SpinalHDL#1273] [Q&A] [answered]
I am trying to create IO ports for a variable number of internal objects.  If the number of objects is 1, I would like to have the input be Bits() type, and if its more than one, I would like a Vec of Bits.
```
case class element() extends Component {
  val io = new Bundle {
    val dataIn = in(Bits(8 bits))
    val dataOut = out(Bits(8 bits))
  }
  ....
}

case class TopWrapper(nElem: Int) extends Component {
  val io = new Bundle {
    val dataIn = if(nElem == 1) in(Bits(8 bits)) else in(Vec(Bits(8 bits)))
    val dataOut = if(nElem == 1) out(Bits(8 bits)) else out(Vec(Bits(8 bits)))
  }

  if (nElem == 1) {
    val elem = new element()

    elem.io.dataIn := io.dataIn
    io.dataOut := elem.io.dataOut
  } else {
    val elems = Array.fill(nElem)(new element())

    for ((elem, idx) <- elems.zipWithIndex) {
      elem.io.dataIn := io.dataIn(idx)
      io.dataOut(idx) := elem.io.dataOut
    }
  }
}
```

But when I try to do this, I get several error messages:

On the nElem == 1 branch of the io assignments:
```
overloaded method := with alternatives:
[error]   (value: String)Unit <and>
[error]   (rangesValue: (Any, Any),_rangesValues: (Any, Any)*)Unit <and>
[error]   (that: spinal.core.Bits)(implicit loc: spinal.idslplugin.Location): Unit
[error]  cannot be applied to (spinal.core.Data{def clone(): this.type})
[error]       elem.io.dataIn := io.dataIn
```

On the else branch of the io assignments:
```
spinal.core.Data{def clone(): this.type} does not take parameters
[error]         io.dataOut(idx) := elem.io.dataOut
```

Is something like this possible, or do I need to seperate the vectorized and non-vectorized IO at the top level into different variables and generate them in a mutually exclusive way, like:

```
val dataIn = (nElem == 1) generate in(Bits(8 bits))
val dataInVec = (nElem > 1) generate in(Vec(Bits(8 bits)))
```

**Comment by mrberman87** [comment]
@Dolu1990 Do you have any suggestions?

**Comment by Dolu1990** [comment, marked as answer]
Hi,

> I need to seperate the vectorized and non-vectorized IO 

I would say you would need it, unless you are willing to add function in the io bundle to provide read write access to them.

But overall, why do you want to separate them ?

Why not just using Vec all the way ? 

> in(Vec(Bits(8 bits)))

Isn't it : in(Vec.fill(nElem)(Bits(8 bits))) ?

**Reply by mrberman87** [reply]
Going Vec all the way I may be easier.  I was thinking about a user's perspective in that a single instance may want it without the Vec, but it's just in how to wire it up and not a bit deal in the end.  Thanks!

**Reply by andreasWallner** [reply]
Maybe as a small explanation to the error: You run into this issue since Scala is strictly typed (the type of a variable must be known at compile time). `Bits` and `Vec` are different types, so scala can only use what they have in common as the type for e.g. `dataIn`. In this case this is `Data` the internal base-type for all signal types, but that is missing all specific functionality.

And in the end I'm not sure you make it easier for the user: they would have to have a special case for n=1 in case they write a component with a parameter.
